docs/core/diagnostics/eventsource.md

```diff
@@ -20,6 +20,17 @@ your own custom events.
 > Many technologies that integrate with EventSource use the terms 'Tracing' and 'Traces' instead of 'Logging' and 'Logs'.
 > The meaning is the same here.
 
+> [!WARNING]
+> Potential Deadlocks with EventSource Callbacks
+>
+> When implementing EventSources, be cautious with invoking lock-acquiring APIs within EventSource callbacks. EventSource instances are initialized early in the runtime, before some core features are fully initialized. As a result, the re-entrant behavior of EventSource callbacks may cause deadlocks if the callback attempts to acquire locks already held by the thread that triggered the callback.
+>
+> To mitigate this risk, consider the following precautions:
```

Comment:
I'd include a recommendation for a re-entrancy guard. For example:

```
[ThreadStatic] t_insideCallback;
public void OnEventWritten(...)
{
    if(t_insideCallback) return; // if our callback triggered the event to occur recursively
                                 // exit now to avoid infinite recursion
    try
    {
        t_insideCallback = true;
        // do whatever callback work
    }
    finally
    {
        t_insideCallback = false;
    }
}
```

I suspect with this in place the TplEventSource case doesn't fail.

Reply:
Added to the list of precautions